### PR TITLE
wgengine/magicsock: fix Conn.Rebind race that let ErrClosed errors be read

### PR DESCRIPTION
**Pull request by bradfitz**
There was a logical race where Conn.Rebind could acquire the
RebindingUDPConn mutex, close the connection, fail to rebind, release
the mutex, and then because the mutex was no longer held, ReceiveIPv4
wouldn't retry reads that failed with net.ErrClosed, letting that
error back to wireguard-go, which would then stop running that receive
IP goroutine.

Instead, keep the RebindingUDPConn mutex held for the entirety of the
replacement in all cases.

Updates tailscale/corp#1289

### CODE REVIEW

**Comment by bradfitz** [comment]
I'll try adding some tests for this too, but wanted to mail this out first.

File: wgengine/magicsock/magicsock.go [hunk 1]
```diff
@@ -2630,28 +2630,38 @@ func (c *Conn) Rebind() {
...
 		c.pconn4.mu.Lock()
+		oldPort := c.pconn4.localAddrLocked().Port
 		if err := c.pconn4.pconn.Close(); err != nil {
 			c.logf("magicsock: link change close failed: %v", err)
 		}
-		packetConn, err := c.listenPacket(listenCtx, "udp4", fmt.Sprintf("%s:%d", host, c.port))
-		if err == nil {
+		packetConn, err := c.listenPacket(listenCtx, "udp4", net.JoinHostPort(host, fmt.Sprint(c.port)))
+		if err != nil {
+			c.logf("magicsock: link change unable to bind fixed port %d: %v, falling back to random port", c.port, err)
```
Comment (crawshaw) [inline wgengine/magicsock/magicsock.go:2642]: Now that we are not always using `c.port`, do we need to update the logic in `determineEndpoints` to use the value of `c.pconn4.localAddrLocked().Port` instead?

**Review by crawshaw — COMMENTED** [review]
Yay!

File: wgengine/magicsock/magicsock.go (hunk 1 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:2642]: No, I think it's still as correct as possible for that heuristic/hack.

But kinda useless now compared to NAT-PMP.

**Review by crawshaw — APPROVED** [review]

**Pull request merged**